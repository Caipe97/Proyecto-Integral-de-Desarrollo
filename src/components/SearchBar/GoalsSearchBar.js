import React, { Component } from 'react';
import './SearchBar.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import DataTable from 'react-data-table-component';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import GoalModal from '../Modal/GoalModal';

const paginacionOpciones={
  rowsPerPageText: 'Filas por Página',
  rangeSeparatorText: 'de',
  selectAllRowsItem: true,
  selectAllRowsItemText: 'Todos',
  
}
const customStyles = {
	header: {
		style: {
			minHeight: '56px',
      background:'#CCDDE2',
      borderRadius:'17px',
      textShadow: '1px 1px black',
		},
	},
	headRow: {
		style: {
      backgroundColor: '#B6E052',
			// borderTopStyle: 'solid',
			// borderTopWidth: '1px',
			// borderTopColor: 'black',
		},
	},
	headCells: {
		style: {
			'&:not(:last-of-type)': {
				// borderRightStyle: 'solid',
				// borderRightWidth: '1px',
				// borderRightColor: 'black',
			},
		},
	},
  pagination: {
		style: {
			backgroundColor:'#B6E052'
		},
	},
  row:{
    pointer:{
      backgroundColor:'blue'
    }
  },
};
const conditionalRowStyles = [
    {
      when: row => row.caloriesPerServing > 0,
      style: {
        backgroundColor: '#B6E052',
        color: 'black',
        '&:hover': {
          cursor: 'pointer',
        },
      },
    },
  ];


class GoalsSearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      busqueda: '',
      goals: [],
      columnas:[],
      cantidadGoals: 0
    };
  }

  onChange = async event => {
    event.persist();
    await this.setState({busqueda: event.target.value});
    this.filtrarElementos();
  }

  asignarColumnas=()=>{
    const columnas = [
      {
        cell:(row) => {
          return(
            <GoalModal goal={row} history={this.props.history} onDeleteGoal={this.props.onDeleteGoal} onUpdateCurrentGoalInState={this.props.onUpdateCurrentGoalInState} />
          )
        },
        ignoreRowClick: true,
        allowOverflow: true,
        button: true
      },
      {
        name: 'Nombre',
        selector: row => row.name,
        sortable: true
      },
      {
        name: 'Fecha',
        // selector: row => row.dateStart.toString().replace('T',' ').substring(0,16),
        selector: row => row.dateStart.toString().replace('T',' ').substring(0,7),
        sortable: true
      }
    ];

    this.setState({columnas: columnas});
  }

  filtrarElementos=()=>{
    let search = this.props.goals.filter(goal => {
      if(goal.name.toLowerCase().includes(this.state.busqueda) || goal.dateStart.includes(this.state.busqueda)){
        return goal;
      } else {
        return '';
      }
    });
    this.setState({goals: search});
  }

  async componentDidMount(){
    // await this.props.onGetAllmeals();
    this.setState({goals: this.props.goals});
    this.asignarColumnas();
  }

  async componentDidUpdate(){
    if(this.props.goals.length !== this.state.cantidadGoals){
      await this.props.onGetGoalsFromUser(this.props.userId)
      this.setState({goals: this.props.goals, cantidadGoals: this.props.goals.length});
      this.asignarColumnas();
    }
  }
  
render(){
  return (
    <div className="table-responsive" style={{backgroundColor:'#B6E052'}}>
      <div className="barraBusqueda" style={{ display:'flex', width: "100%",  justifyContent: 'space-between'}}>
        <h1 style={{color: 'rgb(0, 38, 38)', fontFamily: 'Arial', fontSize: 30}}>Metas</h1>
        <div style={{display: 'flex', alignSelf: 'center'}}>
          <input
            type="text"
            placeholder="Buscar por nombre o fecha"
            className="textField"
            name="busqueda"
            value={this.state.busqueda}
            onChange={this.onChange}
            style={{borderRadius:'13px', width: '88%'}}
            />
          <button type="button" className="btnBuscar" style={{borderRadius:'19px', marginLeft: 5}}>
            <FontAwesomeIcon icon={faSearch} />
          </button>
          <button  type="button" className='btn btn--primary btn--s' style={{ boxShadow: '0px 4px 4px grey', backgroundColor:'#f5f6f7', color: 'black', fontSize: '11px', height: 30, marginLeft: 10}}  onClick={async () => {  this.props.history.push("/goals"); }}>Agregar</button>
        </div>
      </div>
      <DataTable
        columns={this.state.columnas}
        data={this.state.goals}
        //title="Comidas"
        pagination
        paginationComponentOptions={paginacionOpciones}
        fixedHeader
        fixedHeaderScrollHeight="600px"
        noDataComponent={
          <span>No se encontró ninguna meta</span>
        }
        conditionalRowStyles={conditionalRowStyles}
        customStyles={customStyles}
        pointer
      />
    </div>
  );
}
}
export default GoalsSearchBar;