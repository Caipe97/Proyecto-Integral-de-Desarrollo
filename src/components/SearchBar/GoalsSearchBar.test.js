import React from 'react';
import { shallow } from 'enzyme';
import GoalsSearchBar from './GoalsSearchBar';

describe('GoalsSearchBar tests', () => {
    let wrapper;
    const historyMock = { push: jest.fn() };
    const mockPropsSignIn2 = {
        onLogout: jest.fn(),
        state : {
            busquedaNombre: '',
            busquedaCategoria: '',
            alimentos: [],
            categorias: [],
            columnas:[],
            cantidadAlimentos: 0,
            cantidadCategorias: 0
          },
               
          goals:[  { name: 'Plan febrero',dateStart: "2021-09-02T00:00:00.000Z",totalCalories: 3000,objectives: []},
          { name: 'Plan enero',dateStart: "2021-09-01T00:00:00.000Z",totalCalories: 7000,objectives: []},   
        ],
        onAddMeal: jest.fn(),
        onGetMealsFromUser:jest.fn(),
        onDeleteMeal: jest.fn(),
        onGetAllFoods:jest.fn(),
        onAddFoodToCurrentMeal: jest.fn(),
        onRemoveFoodFromCurrentMeal:jest.fn(),
        onLogout: jest.fn(),
        onResetCurrentMeal: jest.fn(),
        onUpdateCurrentMeal:jest.fn(),
        onChangeCurrentMealNameAndDateEaten:jest.fn(),    
        onAddCustomFood:jest.fn(), 
        onEditCustomFood:jest.fn(),
        onDeleteCustomFood:jest.fn(),
        onGetFoodCategories: jest.fn(),
        onCreateCategory: jest.fn(),
        onDeleteCategory: jest.fn(),
        onEditCategory: jest.fn(),
        preventDefault:  jest.fn(),
        onAddObjectiveToCurrentGoal: jest.fn(),
        onAddGoal: jest.fn(),
        onGetGoalsFromUser:jest.fn(),
        onResetCurrentGoal: jest.fn(),
        onChangeCurrentGoalNameTotalCaloriesAndDateStart: jest.fn(),
      
        onRemoveObjectiveFromCurrentGoal: jest.fn(),
        onUpdateCurrentGoal: jest.fn(),
        filtrarELementos:jest.fn()
    };

    beforeEach(() => {
        wrapper = shallow(<GoalsSearchBar {...mockPropsSignIn2} history={historyMock} />);
         
    })

    it('expect to render GoalsSearchBar component to location state ', async() => {
        expect(wrapper).toMatchSnapshot();//if
        await wrapper.instance().componentDidMount();
    })
    it('expect to render GoalsSearchBar component to location state2 ', async() => {
        expect(wrapper).toMatchSnapshot();//if
        await wrapper.instance().componentDidMount();
        await wrapper.instance().componentDidUpdate();
    })
    it('OnChange expect t11', async () => {

        // jest.useFakeTimers();
        await wrapper.find('[type="button"]').at(0).simulate('click',jest.fn);
        
        wrapper.find('[type="button"]').at(0).simulate('click',{ persist: jest.fn()});
        //
        
        //expect(wrapper.find('filtrarElementos').instance()).toHaveBeenCalledTimes(1);
    })
    it('expect to change  onclick buttons', async () => {
       
        await wrapper.find('button').at(0).simulate('click');
       // await wrapper.find('[type="button"]').at(2).simulate('click');
       expect(wrapper.find('div')).toBeDefined();
        expect(wrapper.find('div.p').contains('')).toBeDefined();
      
    })
    it('Agregar should push to goals ', async () => {
        await wrapper.find('button').at(1).simulate('click', { push: jest.fn() });
       // expect(wrapper.instance().props.onRegister).toHaveBeenCalledTimes(1);
        expect(wrapper.instance().props.history.push).toHaveBeenCalledTimes(1);
        expect(historyMock.push.mock.calls[0]).toEqual(['/goals']);
    })
    // it('OnChange should because name contains search value',() => {
    //     wrapper.find('input').at(0).simulate('change', { target: { name: 'busqueda', value: 'enero' } });
    //     //wrapper.instance().filtrarElementos();

    // expect(wrapper.props.filtrarELementos).toHaveBeenCalledTimes(0);
    // })
})
        //wrapper.find('a').simulate('blur', wrapper.find('a').simulate('blur', { persist: jest.fn()}););
