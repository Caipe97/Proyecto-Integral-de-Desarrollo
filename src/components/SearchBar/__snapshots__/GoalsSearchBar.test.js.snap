// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalsSearchBar tests expect to render GoalsSearchBar component to location state  1`] = `
<div
  className="table-responsive"
  style={
    Object {
      "backgroundColor": "#B6E052",
    }
  }
>
  <div
    className="barraBusqueda"
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <h1
      style={
        Object {
          "color": "rgb(0, 38, 38)",
          "fontFamily": "Arial",
          "fontSize": 30,
        }
      }
    >
      Metas
    </h1>
    <div
      style={
        Object {
          "alignSelf": "center",
          "display": "flex",
        }
      }
    >
      <input
        className="textField"
        name="busqueda"
        onChange={[Function]}
        placeholder="Buscar por nombre o fecha"
        style={
          Object {
            "borderRadius": "13px",
            "width": "88%",
          }
        }
        type="text"
        value=""
      />
      <button
        className="btnBuscar"
        style={
          Object {
            "borderRadius": "19px",
            "marginLeft": 5,
          }
        }
        type="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f002",
                "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
              ],
              "iconName": "search",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </button>
      <button
        className="btn btn--primary btn--s"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#f5f6f7",
            "boxShadow": "0px 4px 4px grey",
            "color": "black",
            "fontSize": "11px",
            "height": 30,
            "marginLeft": 10,
          }
        }
        type="button"
      >
        Agregar
      </button>
    </div>
  </div>
  <Memo()
    columns={
      Array [
        Object {
          "allowOverflow": true,
          "button": true,
          "cell": [Function],
          "ignoreRowClick": true,
        },
        Object {
          "name": "Nombre",
          "selector": [Function],
          "sortable": true,
        },
        Object {
          "name": "Fecha",
          "selector": [Function],
          "sortable": true,
        },
      ]
    }
    conditionalRowStyles={
      Array [
        Object {
          "style": Object {
            "&:hover": Object {
              "cursor": "pointer",
            },
            "backgroundColor": "#B6E052",
            "color": "black",
          },
          "when": [Function],
        },
      ]
    }
    customStyles={
      Object {
        "headCells": Object {
          "style": Object {
            "&:not(:last-of-type)": Object {},
          },
        },
        "headRow": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "header": Object {
          "style": Object {
            "background": "#CCDDE2",
            "borderRadius": "17px",
            "minHeight": "56px",
            "textShadow": "1px 1px black",
          },
        },
        "pagination": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "row": Object {
          "pointer": Object {
            "backgroundColor": "blue",
          },
        },
      }
    }
    data={
      Array [
        Object {
          "dateStart": "2021-09-02T00:00:00.000Z",
          "name": "Plan febrero",
          "objectives": Array [],
          "totalCalories": 3000,
        },
        Object {
          "dateStart": "2021-09-01T00:00:00.000Z",
          "name": "Plan enero",
          "objectives": Array [],
          "totalCalories": 7000,
        },
      ]
    }
    fixedHeader={true}
    fixedHeaderScrollHeight="600px"
    noDataComponent={
      <span>
        No se encontró ninguna meta
      </span>
    }
    pagination={true}
    paginationComponentOptions={
      Object {
        "rangeSeparatorText": "de",
        "rowsPerPageText": "Filas por Página",
        "selectAllRowsItem": true,
        "selectAllRowsItemText": "Todos",
      }
    }
    pointer={true}
  />
</div>
`;

exports[`GoalsSearchBar tests expect to render GoalsSearchBar component to location state2  1`] = `
<div
  className="table-responsive"
  style={
    Object {
      "backgroundColor": "#B6E052",
    }
  }
>
  <div
    className="barraBusqueda"
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <h1
      style={
        Object {
          "color": "rgb(0, 38, 38)",
          "fontFamily": "Arial",
          "fontSize": 30,
        }
      }
    >
      Metas
    </h1>
    <div
      style={
        Object {
          "alignSelf": "center",
          "display": "flex",
        }
      }
    >
      <input
        className="textField"
        name="busqueda"
        onChange={[Function]}
        placeholder="Buscar por nombre o fecha"
        style={
          Object {
            "borderRadius": "13px",
            "width": "88%",
          }
        }
        type="text"
        value=""
      />
      <button
        className="btnBuscar"
        style={
          Object {
            "borderRadius": "19px",
            "marginLeft": 5,
          }
        }
        type="button"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f002",
                "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
              ],
              "iconName": "search",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </button>
      <button
        className="btn btn--primary btn--s"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#f5f6f7",
            "boxShadow": "0px 4px 4px grey",
            "color": "black",
            "fontSize": "11px",
            "height": 30,
            "marginLeft": 10,
          }
        }
        type="button"
      >
        Agregar
      </button>
    </div>
  </div>
  <Memo()
    columns={
      Array [
        Object {
          "allowOverflow": true,
          "button": true,
          "cell": [Function],
          "ignoreRowClick": true,
        },
        Object {
          "name": "Nombre",
          "selector": [Function],
          "sortable": true,
        },
        Object {
          "name": "Fecha",
          "selector": [Function],
          "sortable": true,
        },
      ]
    }
    conditionalRowStyles={
      Array [
        Object {
          "style": Object {
            "&:hover": Object {
              "cursor": "pointer",
            },
            "backgroundColor": "#B6E052",
            "color": "black",
          },
          "when": [Function],
        },
      ]
    }
    customStyles={
      Object {
        "headCells": Object {
          "style": Object {
            "&:not(:last-of-type)": Object {},
          },
        },
        "headRow": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "header": Object {
          "style": Object {
            "background": "#CCDDE2",
            "borderRadius": "17px",
            "minHeight": "56px",
            "textShadow": "1px 1px black",
          },
        },
        "pagination": Object {
          "style": Object {
            "backgroundColor": "#B6E052",
          },
        },
        "row": Object {
          "pointer": Object {
            "backgroundColor": "blue",
          },
        },
      }
    }
    data={
      Array [
        Object {
          "dateStart": "2021-09-02T00:00:00.000Z",
          "name": "Plan febrero",
          "objectives": Array [],
          "totalCalories": 3000,
        },
        Object {
          "dateStart": "2021-09-01T00:00:00.000Z",
          "name": "Plan enero",
          "objectives": Array [],
          "totalCalories": 7000,
        },
      ]
    }
    fixedHeader={true}
    fixedHeaderScrollHeight="600px"
    noDataComponent={
      <span>
        No se encontró ninguna meta
      </span>
    }
    pagination={true}
    paginationComponentOptions={
      Object {
        "rangeSeparatorText": "de",
        "rowsPerPageText": "Filas por Página",
        "selectAllRowsItem": true,
        "selectAllRowsItemText": "Todos",
      }
    }
    pointer={true}
  />
</div>
`;
